File: Team 3/Python/LRUCache.py
class LRUCache:
    
    def __init__(self, capacity: int):
        # Declaramos el dictionario para los valores y la capacidad total de la cache
        self.address = OrderedDict()
        self.values = capacity

    def get(self, key: int) -> int:
        if key not in self.address:
            return -1
        else:
            # Al momento de leer el valor se vuelve el mas usado, por lo que lo movemos hasta adelante
            self.address.move_to_end(key, last=False)
            return self.address[key]

    def put(self, key: int, value: int) -> None:
        if key in self.address:
            # Hacemos update del valor que ya se tiene ademas de que lo movemos al principio
            self.address.move_to_end(key, last=False)
        else:
            if len(self.address) >= self.values:
                # Quitamos el dato menos ocupado en caso de que se quiera agregar un valor mas de la capacidad
                self.address.popitem(last=True)
        
        self.address[key] = value
        self.address.move_to_end(key, last=False)

File: Team 3/Typescript/LRUCache.ts
class LRUCache {
    private address: Map<number, { key: number, value: number }>;
    private values: number;

    // Creamos la cache y declaramos la capacidad
    constructor(capacity: number) {
        this.address = new Map();
        this.values = capacity;
    }

    get(key: number): number {
        if (!this.address.has(key)) {
            return -1;
        }
        
        const { value } = this.address.get(key)!;
        // Movemos hasta adelante el valor recientemente leido
        this.address.delete(key);
        this.address.set(key, { key, value });
        
        return value;
    }

    put(key: number, value: number): void {
        if (this.address.has(key)) {
            // Eliminamos del map el valor que vamos a poner ya que vamos a hacer update
            this.address.delete(key);
            this.address.set(key, { key, value });
        } else {
            if (this.address.size >= this.values) {
                // En caso de que al momento de insertar un nuevo valor sea mayor, eliminamos el menos usado
                const oldestKey = this.address.keys().next().value;
                this.address.delete(oldestKey);
            }

            this.address.set(key, { key, value });
        }
    }
}